modules/api/php/endpoints/candidate/visit/visit_0_0_4_dev.class.inc

```diff
@@ -0,0 +1,155 @@
+ * @author   Xavier Lecours Boucher <xavier.lecours@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+class Visit_0_0_4_Dev extends Visit
+{
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return [
+            'GET',
+            'PATCH',
+            'PUT',
+        ];
+    }
+
+    /**
+     * Versions of the LORIS API which are supported by this
+     * endpoint.
+     *
+     * @return array a list of supported API versions.
+     */
+    protected function supportedVersions() : array
+    {
+        return [
+            'v0.0.4-dev'
+        ];
+    }
+
+    /**
+     * Handles a PATCH request.
+     * For now, only support a NextStageDate argument to start next stage.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    protected function handlePATCH(
+        ServerRequestInterface $request
+    ): ResponseInterface {
+        $data          = json_decode((string) $request->getBody(), true);
+        $nextStageDate = $data['NextStageDate'] ?? null;
+
+        if ($this->_visit === null) {
+            return new \LORIS\Http\Response\JSON\NotFound('Visit not found');
+        }
+
+        $response = $this->validatePUT($request);
+        if ($response !== null) {
+            return $response;
+        }
+
+        if (!isset($data['Visit'])
+            || $data['Visit'] !== $this->_visit->getVisitLabel()
+        ) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'The Visit field does not match this visit'
+            );
+        }
+
+        if (!$nextStageDate) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'There is no stage date specified'
+            );
+        }
+
+        $date_format = 'Y-m-d';
+        $date        = \DateTime::createFromFormat($date_format, $nextStageDate);
+
+        if (!($date && $date->format($date_format) === $nextStageDate)) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'The date specified is invalid. Date must be of the form YYYY-MM-DD'
+            );
+        }
+
+        if ($nextStageDate > date($date_format)) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'The date specified must be not later than today'
+            );
+        }
+
+        if ($this->_visit->getCurrentStage() !== 'Not Started') {
+            return new \LORIS\Http\Response\JSON\Conflict(
+                'This visit is already started'
+            );
+        }
+
+        $newStage = $this->_visit->getNextStage();
+        if (!$newStage) {
+            return new \LORIS\Http\Response\JSON\Conflict(
+                "Next stage is null"
+            );
+        }
+
```

Comment:
The only validations missing are useScanDone and allowPrenatalTimepoints config dependent ones. see: https://github.com/aces/Loris/blob/main/modules/next_stage/php/next_stage.class.inc#L256-L273

@driusan All this, and the code in the next_stage module should all be migrate to the timepoint class IMO. It would be convinient to call `$visit->startNextState(stage='Visit', startdate='2020-04-23', scandone=false);`

Reply:
It is not in the scope of this PR, i know :)